core/arch/arm/plat-stm32mp1/plat_huk.c

```diff
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2022, Linaro Limited
+ * Copyright (c) 2022, Foundries.io Limited
+ */
+
+#include <assert.h>
+#include <config.h>
+#include <crypto/crypto.h>
+#include <drivers/stm32_bsec.h>
+#include <kernel/tee_common_otp.h>
+#include <platform_config.h>
+#include <string.h>
+
+#if !defined(CFG_STM32MP1_HUK_BSEC_BASE)
+#error "CFG_STM32MP1_HUK_BSEC_BASE undefined"
+#endif
+
+static struct {
+	uint8_t key[HW_UNIQUE_KEY_LENGTH];
+	bool ready;
+} huk;
+
+TEE_Result tee_otp_get_hw_unique_key(struct tee_hw_unique_key *hwkey)
+{
+	const uint32_t bsec_base = CFG_STM32MP1_HUK_BSEC_BASE;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint8_t key[16] = { 0 };
+	uint8_t tag[16] = { 0 };
+	uint8_t uid[16] = { 0 };
+	uint8_t nce[12] = {
+		0xd2, 0x45, 0x0e, 0x07, 0xea, 0x5d,
+		0xe0, 0x42, 0x6c, 0x0f, 0xa1, 0x33,
+	};
+	size_t enc_len = sizeof(huk.key);
+	size_t tag_len = sizeof(tag);
+	size_t uid_len = sizeof(uid);
+	uint32_t word = 0;
+	void *ctx = NULL;
+	size_t n = 0;
+
+	if (huk.ready)
+		goto out;
+
+	for (n = 0; n < sizeof(key) / sizeof(uint32_t); n++) {
+		res = stm32_bsec_shadow_register(bsec_base + n);
+		if (res)
+			return res;
+
+		if (IS_ENABLED(CFG_STM32MP1_HUK_BSEC_SHADOW_TESTKEY)) {
+			res = stm32_bsec_write_otp(0, bsec_base + n);
+			if (res)
+				return res;
+		}
+
+		res = stm32_bsec_read_otp(&word, bsec_base + n);
+		if (res)
+			return res;
+
+		memcpy(key + n * sizeof(uint32_t), &word, sizeof(uint32_t));
+	}
+
+	for (n = 0; n < (UID2_OTP - UID0_OTP + 1); n++) {
+		res = stm32_bsec_read_otp(&word, UID0_OTP + n);
+		if (res)
+			return res;
+
+		memcpy(uid + n * sizeof(uint32_t), &word, sizeof(uint32_t));
+	}
+
+	/* HUK becomes the AES-GCM encrypted UID using a 128 bit key from OTP */
+	res = crypto_authenc_alloc_ctx(&ctx, TEE_ALG_AES_CCM);
+	if (res)
+		return res;
+
+	res = crypto_authenc_init(ctx, TEE_MODE_ENCRYPT, key, sizeof(key), nce,
+				  sizeof(nce), TEE_AES_BLOCK_SIZE, 0, id_len);
+	if (res)
+		return res;
+
+	res = crypto_authenc_enc_final(ctx, uid, sizeof(uid), huk.key, &enc_len,
+				       tag, &tag_len);
+	if (res)
+		return res;
+
+	crypto_authenc_final(ctx);
+	crypto_authenc_free_ctx(ctx);
+
+	huk.ready = true;
+
+	if (!stm32mp_is_closed_device())
```

Comment:
To ensure key don't change, we should test the permanent-lock of each read OTP:
```patch
+	huk_fuses_locked = false;
 	for (n = 0; n < sizeof(key) / sizeof(uint32_t); n++) {
+		boot lock = false;
+
+		res = stm32_bsec_read_permanent_lock(bsec_base + n, &lock);
+		if (res)
+			return res;
+		if (!lock)
+			huk_fuses_locked = false;	
+
 		res = stm32_bsec_shadow_register(bsec_base + n);
 		if (res)
 			return res;
```

The test (at line 91) should also depend on CFG_STM32MP1_HUK_BSEC_SHADOW_TESTKEY.